swarm/pss/pss_test.go

```diff
@@ -475,6 +482,196 @@ func testAsymSend(t *testing.T) {
+	msgcount, _ := strconv.ParseInt(paramstring[2], 10, 0)
+	addrsize, _ := strconv.ParseInt(paramstring[3], 10, 0)
+	messagedelayvarianceusec := (int(msgcount)/1000 + 1) * 1000 * 1000
+	log.Info("network test", "nodecount", nodecount, "msgcount", msgcount, "addrhintsize", addrsize, "sendtimevariance", messagedelayvarianceusec/(1000*1000))
+
+	nodes := make([]discover.NodeID, nodecount)
+	bzzaddrs := make(map[discover.NodeID][]byte, nodecount)
+	rpcs := make(map[discover.NodeID]*rpc.Client, nodecount)
+	pubkeys := make(map[discover.NodeID][]byte, nodecount)
+
+	sentmsgs := make([][]byte, msgcount)
+	recvmsgs := make([]bool, msgcount)
+	nodemsgcount := make(map[discover.NodeID]int, nodecount)
+
+	trigger := make(chan discover.NodeID)
+
+	var adapter adapters.NodeAdapter
+	if paramstring[4] == "exec" {
+		dirname, err := ioutil.TempDir(".", "")
+		if err != nil {
+			t.Fatal(err)
+		}
+		adapter = adapters.NewExecAdapter(dirname)
+	} else {
+		adapter = adapters.NewSocketAdapter(services)
+	}
+	net := simulations.NewNetwork(adapter, &simulations.NetworkConfig{
+		ID: "0",
+	})
+	defer net.Shutdown()
+
+	f, err := os.Open(fmt.Sprintf("testdata/snapshot_%s.json", paramstring[1]))
+	if err != nil {
+		t.Fatal(err)
+	}
+	jsonbyte, err := ioutil.ReadAll(f)
+	if err != nil {
+		t.Fatal(err)
+	}
+	var snap simulations.Snapshot
+	err = json.Unmarshal(jsonbyte, &snap)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = net.Load(&snap)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	triggerChecks := func(trigger chan discover.NodeID, id discover.NodeID, rpcclient *rpc.Client) error {
+		msgC := make(chan APIMsg)
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+		defer cancel()
+		sub, err := rpcclient.Subscribe(ctx, "pss", msgC, "receive", hextopic)
+		if err != nil {
+			t.Fatal(err)
+		}
+		go func() {
+			defer sub.Unsubscribe()
+			for {
+				select {
+				case recvmsg := <-msgC:
+					idx, _ := binary.Uvarint(recvmsg.Msg)
+					if recvmsgs[idx] == false {
+						log.Debug("msg recv", "idx", idx, "id", id)
+						recvmsgs[idx] = true
+						trigger <- id
+					}
+				case <-sub.Err():
+					return
+				}
+			}
+		}()
+		return nil
+	}
+
+	for i, nod := range net.GetNodes() {
+		nodes[i] = nod.ID()
+		rpcs[nodes[i]], err = nod.Client()
+		if err != nil {
+			t.Fatal(err)
+		}
+		pubkey := make([]byte, 65)
+		pubkeys[nod.ID()] = make([]byte, 65)
+		err = rpcs[nodes[i]].Call(&pubkey, "pss_getPublicKey")
+		if err != nil {
+			t.Fatal(err)
+		}
+		copy(pubkeys[nod.ID()], pubkey)
+		addr := make([]byte, 32)
+		err = rpcs[nodes[i]].Call(&addr, "pss_baseAddr")
+		if err != nil {
+			t.Fatal(err)
+		}
+		bzzaddrs[nodes[i]] = addr[:addrsize]
+		err = triggerChecks(trigger, nodes[i], rpcs[nodes[i]])
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	messagedelayhigh := 0
+	for i := 0; i < int(msgcount); i++ {
+		sendnodeidx := rand.Intn(int(nodecount))
+		recvnodeidx := rand.Intn(int(nodecount - 1))
+		if recvnodeidx >= sendnodeidx {
+			recvnodeidx++
```

Comment:
Why does this get incremented when the receiving node index is greater than the sending node index?

Reply:
random index for sender and receiver should not be same. It's to ensure randomness of the correct set.

Reply:
I understand we don't want them to be the same, but let's say `recvnodeidx = 10` and `sendnodeidx = 2`, why would you then need to increment `recvnodeidx` by 1 to 11?

Reply:
We choose send s from [0,n>

We choose recv r from [0,n-1>

But since r is in effect [0,s>, <s,n>, we simply increment if it should happen to be s.

What change do you propose?